src/modules/rtpengine/rtpengine.c

```diff
@@ -1240,8 +1250,208 @@ static int rtpengine_add_rtpengine_set(char *rtp_proxies, unsigned int weight,
+{
+	char * p;
+	p = (char* )val;
+
+	if(p==0 || *p=='\0'){
+		return 0;
+	}
+
+	rtpengine_dtmf_event_sock.s = p;
+	rtpengine_dtmf_event_sock.len = strlen(rtpengine_dtmf_event_sock.s);
+
+	return 0;
+}
 
-static int fixup_set_id(void **param, int param_no)
+static void rtpengine_dtmf_events_loop(int rank)
+{
+	int ret;
+	char *p;
+	str s_port;
+	unsigned int port;
+	union sockaddr_union udp_addr;
+	char buffer[RTPENGINE_DTMF_EVENT_BUFFER];
+
+	p = q_memchr(rtpengine_dtmf_event_sock.s, ':', rtpengine_dtmf_event_sock.len);
+	if (!p) {
+		LM_ERR("failed to initialize dtmf event listener because no port was specified %.*s!\n", rtpengine_dtmf_event_sock.len, rtpengine_dtmf_event_sock.s);
+		return;
+	}
+
+	s_port.s = p + 1;
+	s_port.len = rtpengine_dtmf_event_sock.s + rtpengine_dtmf_event_sock.len - s_port.s;
+
+	if (s_port.len <= 0 || str2int(&s_port, &port) < 0 || port > 65535) {
+		LM_ERR("failed to initialize dtmf event listener because port is invalid %.*s\n", rtpengine_dtmf_event_sock.len, rtpengine_dtmf_event_sock.s);
+		return;
+	}
+	rtpengine_dtmf_event_sock.len -= s_port.len + 1;
+	trim(&rtpengine_dtmf_event_sock);
+	rtpengine_dtmf_event_sock.s[rtpengine_dtmf_event_sock.len] = '\0';
+
+	memset(&udp_addr, 0, sizeof(udp_addr));
+
+	if (rtpengine_dtmf_event_sock.s[0] == '[') {
+		udp_addr.sin6.sin6_family = AF_INET6;
+		udp_addr.sin6.sin6_port = htons(port);
+		ret = inet_pton(AF_INET, rtpengine_dtmf_event_sock.s, &udp_addr.sin6.sin6_addr);
+	} else {
+		udp_addr.sin.sin_family = AF_INET;
+		udp_addr.sin.sin_port = htons(port);
+		ret = inet_pton(AF_INET, rtpengine_dtmf_event_sock.s, &udp_addr.sin.sin_addr);
+	}			
+
+	if (ret != 1) {
+		LM_ERR("failed to initialize dtmf event listener because address could not be created for %s\n", rtpengine_dtmf_event_sock.s);
+		return;
+	}
+
+	rtpengine_dtmf_event_fd = socket(udp_addr.s.sa_family, SOCK_DGRAM, 0);
+
+	if(rtpengine_dtmf_event_fd < 0) {
+		LM_ERR("can't create socket\n");
+		return;
+	}
+	
+	if (bind(rtpengine_dtmf_event_fd, (struct sockaddr*)&udp_addr.s, sizeof(udp_addr.s)) < 0) {
+		LM_ERR("could not bind dtmf events socket %s:%u (%s:%d)\n", rtpengine_dtmf_event_sock.s, port, strerror(errno), errno);
+		goto end;
+	}
+	
+	LM_INFO("dtmf event listener started on %s:%u\n", rtpengine_dtmf_event_sock.s, port);
+	
+	for (;;) {		
+		do
+			ret = read(rtpengine_dtmf_event_fd, buffer, RTPENGINE_DTMF_EVENT_BUFFER);			
+		while (ret == -1 && errno == EINTR);
+
+		if (ret < 0) {
+			LM_ERR("problem reading on socket %s:%u (%s:%d)\n", rtpengine_dtmf_event_sock.s, port, strerror(errno), errno);
+			goto end;
+		}
+
+		if (dtmf_event_rt == -1) {
+			LM_NOTICE("nothing to do - nobody is listening!\n");
+			goto end;
+		}
+
+		p = shm_malloc(ret + 1);
+		if (!p) {
+			LM_ERR("could not allocate %d for buffer %.*s\n", ret, ret, buffer);
+			goto end;
+		}
+		memcpy(p, buffer, ret);
+		p[ret] = '\0';		
+
+		if (rtpengine_raise_dtmf_event(p, ret) < 0) {
```

Comment:
I think this leaks the `p` buffer if the return value == 0

Reply:
Fixed.